RavenDB-18946 : Sharding - cloud backup and restore  + encrypted backup and restore

aviv86 opened the pull request:
### Issue link

- https://issues.hibernatingrhinos.com/issue/RavenDB-18206/Sharding-Restore-Import
- https://issues.hibernatingrhinos.com/issue/RavenDB-18946/Sharding-Backup-Test-cloud-backup-and-encrypted-backup

### Additional description

- Sharded backup and restore from cloud 
- Sharded encrypted backup and restore
- Tests

### Type of change

- New feature

### How risky is the change?

- Low 

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

aviv86 committed:
RavenDB-18206 : sharded restore from cloud backup

aviv86 committed:
RavenDB-18946 : test encrypted backup and restore of sharded database

karmeli87 commented on src/Raven.Server/Documents/PeriodicBackup/Restore/AbstractRestoreBackupTask.cs:
```diff
@@ -106,6 +105,9 @@ public async Task<IOperationResult> Execute(Action<IOperationProgress> onProgres
 
         protected abstract Task RestoreAsync();
 
+        protected virtual void PutSecretKey() => 
```
let's have single impl for `PutSecretKey(string databaseName)`

aviv86 replied:
done

aviv86 committed:
RavenDB-18206 : address PR comment

aviv86 committed:
RavenDB-18206 : bug fixes
- SingleShardRestoreBackupTask needs to save database record after restore
- GetGoogleCloudSettings should not override static GoogleCloudFactAttribute.GoogleCloudSettings

karmeli87 commented on src/Raven.Server/Documents/PeriodicBackup/Restore/Sharding/SingleShardRestoreBackupTask.cs:
```diff
@@ -44,7 +40,7 @@ protected override Task OnAfterRestoreAsync()
             SmugglerBase.EnsureProcessed(Result, skipped: false);
             Progress.Invoke(Result.Progress);
-            return Task.CompletedTask;
+            return SaveDatabaseRecordAsync(RestoreSettings.DatabaseRecord.DatabaseName, RestoreSettings.DatabaseRecord, databaseValues: null, Result, Progress);
```
why do we need to save the database record here?

aviv86 replied:
on import I'm using `onDatabaseRecordAction` ([here](
https://github.com/ravendb/ravendb/blob/v6.0/src/Raven.Server/Documents/PeriodicBackup/Restore/AbstractRestoreBackupTask.cs#L260)) - in order to enable revisions from smuggler record before import, like we do in regular restore.
If we don't save the database record afterwards then it's basically like skipping database-record processing

in [RavenDB-19202](https://issues.hibernatingrhinos.com/issue/RavenDB-19202) we'll make adjustments such that the shard with the most up-to-date database record will be the only one that operates on database record

ppekrol commented on src/Raven.Client/Documents/Operations/Backups/Sharding/ShardRestoreSettings.cs:
```diff
@@ -38,7 +38,7 @@ public class SingleShardRestoreSetting : IDynamicJson
         public int ShardNumber { get; set; }
         public string NodeTag { get; set; }
-        public string BackupPath { get; set; }
+        public string FolderName { get; set; }
```
this cannot point to a single file, correct?

aviv86 replied:
actually I think it will also work if it points to a single file.. at least on google cloud it does (didn't check on s3 or azure)
same thing in regular restore, we can have `GoogleCloudSettings.RemoteFolderName` pointing to a single file and it will work

you think I should revert back to `BackupPath` ?

ppekrol replied:
I'm merging this PR, @aviv86 I would like to have consistent naming, what it should be?
